### PR TITLE
baremetal: Enable Ironic on the bootstrap VM

### PR DESCRIPTION
**Pull request by hardys**
This series enables Ironic on the bootstrap VM (Issue #2060), based on work first started via  https://github.com/openshift-metal3/kni-installer/pull/100

Since this also reworks our image handling to use the internal image references, it also completes issue #2064 and I think may complete issue #2037

Note this depends on https://github.com/openshift/installer/pull/2061

### CODE REVIEW

**Comment by hardys** [comment]
Note - I'm not sure if the increase to 6G is still needed, I added it when I was having issues with libpod (since resolved via https://github.com/openshift/installer/pull/1941) - I need to re-test with the default memory and prove everything still works, if so I'll drop that patch from the series.

**Comment by hardys** [comment]
> /label platform/baremetal

Note that the metal3 CI results on this PR won't be valid, because we need corresponding dev-scripts changes, I'll test it via https://github.com/openshift-metal3/dev-scripts/pull/642 when it's fully working locally (still debugging some issues atm).

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 1]
```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+
+set -ex
+
+IRONIC_IMAGE=${IRONIC_IMAGE:-"quay.io/metal3-io/ironic:master"}
+IRONIC_INSPECTOR_IMAGE=${IRONIC_INSPECTOR_IMAGE:-"quay.io/metal3-io/ironic-inspector:master"}
+IPA_DOWNLOADER_IMAGE=${IPA_DOWNLOADER_IMAGE:-"quay.io/metal3-io/ironic-ipa-downloader:master"}
+COREOS_DOWNLOADER_IMAGE=${COREOS_DOWNLOADER_IMAGE:-"quay.io/openshift-metal3/rhcos-downloader:master"}
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Maybe add a comment here noting that all of these will be changed to get image specs from the release image, once all of these images are included there?  We should link to a bug tracking that.

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+# This image is templated in via the installer pkg/asset/ignition/bootstrap/bootstrap.go
+RHCOS_IMAGE_BASE_URL="{{.BootImage}}"
+
+# First we stop any previously started containers, because ExecStop only runs when the ExecStart process
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: It sounds like this comment wasn't finished - "only runs when the ExecStart process ..."

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+# e.g this script is still running, but we exit if *any* of the containers exits unexpectedly
+for name in ironic-api ironic-conductor ironic-inspector dnsmasq httpd mariadb ipa-downloader coreos-downloader; do
+    podman ps | grep -w "$name$" && podman kill $name
+    podman ps --all | grep -w "$name$" && podman rm $name -f
+done
+
+# Start the provisioning nic if not already started
+# FIXME we should detect the nic in the ironic container (currently relies on the ifname, and won't pick 
+# up the description added here)
+# FIXME The IP/subnet and gateway should be provided via survey input?
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: We're not supporting survey input right now, but support via install-config makes sense.  If not specified in install-config, we'd just use the default (and not any survey input)

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+# FIXME: the ironic containers currently bind to all interfaces, which is not secure in the
+# case of the cluster, but we'll need the installer to access the public nic on the boostrap node
+# so probably we'll need a forwarding rule after https://github.com/metal3-io/ironic-image/pull/56
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: I'm not clear on what the fix is here - change ironic to only bind to the provisioning interface?  ensure the installer is only using that IP?

Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Ah thanks for the reminder, yeah the fix is to bind only on the provisioning nic, but I couldn't do that previously because terraform was accessing ironic on the host via the baremetal bridge - I'll re-test with https://github.com/metal3-io/ironic-image/pull/56 locally as I think it should work now that this series switches terraform to reference the BootstrapProvisioningIP.

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 1 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Ack, good point, I'll raise an issue similar to #2067 to track this

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/929/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/930/

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+PROVISIONING_NIC=ens4
+if ! nmcli -t device | grep "$PROVISIONING_NIC:ethernet:connected:provisioning"; then
+    nmcli c add type ethernet ifname $PROVISIONING_NIC con-name provisioning ip4 172.22.0.2/24 gw4 172.22.0.1
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: want to add a FIXME comment and a follow-up bug about making this subnet configurable in install-config?  I think having it hard coded is fine for 4.2 IMO

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+# Wait for the interface to come up
+# This is how the ironic container currently detects IRONIC_IP, this could probably be improved by using
+# nmcli show provisioning there instead, but we need to confirm that works with the static-ip-manager
+while [ -z "$(ip -4 address show dev "$PROVISIONING_NIC" | grep -oP '(?<=inet\s)\d+(\.\d+){3}' | head -n 1)" ]; do
+    sleep 1
+done
+
+# set password for mariadb
+mariadb_password=$(echo $(date;hostname)|sha256sum |cut -c-20)
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: guess it doesn't matter much, but i see that rhcos has `uuidgen` if you want something more random

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+IRONIC_SHARED_VOLUME="ironic"
+podman volume create $IRONIC_SHARED_VOLUME || true
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Why are errors ignored here?  in case it already exists?

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+# Start the conductor so db sync can happen while images download
+sudo podman run -d --net host --privileged --name ironic-conductor \
+     --env MARIADB_PASSWORD=$mariadb_password \
+     --env PROVISIONING_INTERFACE=$PROVISIONING_NIC \
+     --env OS_CONDUCTOR__HEARTBEAT_TIMEOUT=120 \
+     --entrypoint /bin/runironic-conductor \
+     -v $IRONIC_SHARED_VOLUME:/shared:z ${IRONIC_IMAGE}
+
+# Wait for images to be downloaded/ready
+while ! curl --fail http://localhost:80/images/rhcos-ootpa-latest.qcow2.md5sum ; do sleep 1 ; done
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: I suppose we could detect a failure a bit quicker here if we watched the state of the `coreos-downloader` container, and then did a one time check once the downloader has exited.

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+     --env PROVISIONING_INTERFACE=$PROVISIONING_NIC \
+     -v $IRONIC_SHARED_VOLUME:/shared:z "${IRONIC_INSPECTOR_IMAGE}"
+
+sudo podman run -d --net host --privileged --name ironic-api \
+     --env MARIADB_PASSWORD=$mariadb_password \
+     --env PROVISIONING_INTERFACE=$PROVISIONING_NIC \
+     --entrypoint /bin/runironic-api \
+     -v $IRONIC_SHARED_VOLUME:/shared:z ${IRONIC_IMAGE}
+
+while true; do
```
Comment (russellb) [inline data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Could you add a comment about what this loop is for?

File: pkg/asset/cluster/baremetal/baremetal.go [hunk 10]
```diff
@@ -10,7 +10,8 @@ import (
 // Metadata converts an install configuration to bare metal metadata.
 func Metadata(config *types.InstallConfig) *baremetal.Metadata {
 	return &baremetal.Metadata{
-		LibvirtURI: config.Platform.BareMetal.LibvirtURI,
-		IronicURI:  config.Platform.BareMetal.IronicURI,
+		LibvirtURI:              config.Platform.BareMetal.LibvirtURI,
+		BootstrapProvisioningIP: config.Platform.BareMetal.BootstrapProvisioningIP,
+		ClusterProvisioningIP:      config.Platform.BareMetal.ClusterProvisioningIP,
```
Comment (russellb) [inline pkg/asset/cluster/baremetal/baremetal.go]: looks like there's an alignment issue here

File: pkg/types/baremetal/platform.go [hunk 11]
```diff
@@ -33,10 +24,16 @@ type Platform struct {
...
+	// network for the host running the installer, can be used to cache some
+	// downloaded content e.g RHCOS/IPA images
 	// +optional
-	IronicURI string `json:"ironicURI,omitempty"`
+	ClusterProvisioningIP string `json:"provisioningHostIP,omitempty"`
+
+	// BootstrapProvisioningIP is the IP used on the bootstrap VM to
+	// bring up provisioning services.
+	// +optional
+	BootstrapProvisioningIP string `json:"bootstrapProvisioningIP,omitempty"`
```
Comment (russellb) [inline pkg/types/baremetal/platform.go]: Can you add docs for the new fields to `docs/user/metal/install_ipi.md` ?

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 6 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Ack thanks, this was just copied from dev-scripts but I'll check out uuidgen

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 7 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Yeah so that if one of the service containers exits, which will cause a systemd restart of all-the-things, we'll connect to the existing volume and potentially reuse existing images vs downloading again.

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 8 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Ah yeah that's a good idea, thanks!

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 9 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Yes will do - it's just so the service remains active and we'll restart everything should one of the containers exit unexpectedly.  The alternative would be `RemainAfterExit=yes` but then we lose the ability to restart if something crashes.

File: pkg/types/baremetal/platform.go (hunk 11 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: ++ will do, thanks for the review!

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 1 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: https://github.com/openshift/installer/issues/2090 raised

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 2 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: The comment continues on the following line:

```
# First we stop any previously started containers, because ExecStop only runs when the ExecStart process
# e.g this script is still running, but we exit if *any* of the containers exits unexpectedly

```

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 3 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: I removed all the FIXME comments for now, and added a reference to https://github.com/openshift/installer/issues/2091

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 4 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: I removed the FIXME comments for now, added a link to https://github.com/openshift/installer/issues/2091 and will follow-up wrt the bind IP question via update of the ironic image (which can only really happen after this initial integration lands I think).

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 5 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Added reference to https://github.com/openshift/installer/issues/2091

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 6 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: Done

File: data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template (hunk 8 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/files/usr/local/bin/startironic.sh.template]: looks like `podman wait` will help with this.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/938/

File: pkg/types/baremetal/platform.go (hunk 11 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: Hmm looking at that file, there's only an example install-config, so perhaps it's best to not add these yet, since there will be more required before we can fully support alternate subnets ref  #2091

I will remove the part about requiring an external Ironic tho :)

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/940/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/942/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/949/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/983/

**Comment by metal3ci** [comment]
Build ABORTED, see build http://10.8.144.11:8080/job/dev-tools/986/

File: data/data/baremetal/main.tf [hunk 12]
```diff
@@ -3,8 +3,9 @@ provider "libvirt" {
 }
 
 provider "ironic" {
-  url          = var.ironic_uri
+  url          = "http://${var.bootstrap_provisioning_ip}:6385/v1"
   microversion = "1.52"
+  timeout      = 1200
```
Comment (stbenjam) [inline data/data/baremetal/main.tf]: I've had to bump this up to 1200, which I think is going to cause issues with timeouts.  My connection in RDU took about 15 minutes for the images to finish downloading and Ironic to become available. This may be why CI has been failing on the dev-scripts PR.

**Comment by metal3ci** [comment]
Build ABORTED, see build http://10.8.144.11:8080/job/dev-tools/990/

**Comment by abhinavdahiya** [comment]
> https://github.com/openshift/installer/pull/2079/commits/e03190ed8cff85313b76943c1aca36979783a5b7

Is this required for a dev purpose or is this valid for production envs (CI counts too) ??

**Comment by abhinavdahiya** [comment]
> > [e03190e](https://github.com/openshift/installer/commit/e03190ed8cff85313b76943c1aca36979783a5b7)
> 
> Is this required for a dev purpose or is this valid for production envs (CI counts too) ??

And can you provide a detailed timeline of the steps that cause "api to come up after create resources completes" to exceed 30 mins..

**Comment by stbenjam** [comment]
> > [e03190e](https://github.com/openshift/installer/commit/e03190ed8cff85313b76943c1aca36979783a5b7)
> 
> Is this required for a dev purpose or is this valid for production envs (CI counts too) ??

This is still very much a WIP while I investigate why things are failing in my env in one of RH's data centers - my connection is very slow there.

However,  I don't believe e03190e is needed, because `waitForBootstrapComplete` isn't run until terraform completes, so the relevant time out was our own in terraform-provider-ironic, not the installer's.

It probably helps to spell out what's going on here:

1. Installer creates the 'cluster' asset which kicks off terraform to create bootstrap and masters
2. Terraform starts, and creates the bootstrap resource. The masters aren't created yet, as terraform-provider-ironic is waiting on the API to become available.
3. Bootstrap is provisioned, and runs the `ironic` service which does a bunch of things: namely downloading the images required to provision the bare metal hosts.  The RHCOS image is 650MB, so this takes quite a while on slow connections.
4. When download has finished, Ironic itself gets started
5. Terraform-provider-ironic notices the API becomes available, and provisions the masters

File: data/data/baremetal/main.tf (hunk 12 above)
Comment (stbenjam) [reply data/data/baremetal/main.tf]: I've actually set it to 3600, but this doesn't cause any timeout issue with the installer.  The time out only begins after terraform stands up the infrastructure for bootstrap and masters.

**Comment by stbenjam** [comment]
/assign @abhinavdahiya 

PTAL :-) This is ready for a review, it's [passing our CI](https://github.com/openshift-metal3/dev-scripts/pull/642).

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1018/

**Comment by russellb** [comment]
I think this looks good to go in, and we cancontinue to iterate on it in follow up PRs

/lgtm

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1021/

**Comment by russellb** [comment]
/test e2e-azure

Failed due to a quota issue

File: data/data/bootstrap/baremetal/systemd/units/ironic.service [hunk 13]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Baremetal Deployment Ironic Services
+Wants=network-online.target crio.service
+After=network-online.target crio.service
+
+[Service]
+Type=exec
+ExecStart=/usr/local/bin/startironic.sh
+ExecStop=/usr/local/bin/stopironic.sh
+ConditionPathExists=!/etc/pivot/image-pullspec
```
Comment (abhinavdahiya) [inline data/data/bootstrap/baremetal/systemd/units/ironic.service]: what's the significance of this condition? Why is this required?

File: pkg/asset/cluster/baremetal/baremetal.go [hunk 14]
```diff
@@ -10,7 +10,8 @@ import (
 // Metadata converts an install configuration to bare metal metadata.
 func Metadata(config *types.InstallConfig) *baremetal.Metadata {
 	return &baremetal.Metadata{
-		LibvirtURI: config.Platform.BareMetal.LibvirtURI,
-		IronicURI:  config.Platform.BareMetal.IronicURI,
+		LibvirtURI:              config.Platform.BareMetal.LibvirtURI,
+		BootstrapProvisioningIP: config.Platform.BareMetal.BootstrapProvisioningIP,
```
Comment (abhinavdahiya) [inline pkg/asset/cluster/baremetal/baremetal.go]: why is the `BootstrapProvisioningIP` and `ClusterProvisioningIP` required in the metadata?

File: pkg/asset/machines/baremetal/machines.go [hunk 15]
```diff
@@ -60,11 +62,18 @@ func Machines(clusterID string, config *types.InstallConfig, pool *types.Machine
 	return machines, nil
 }
 
-func provider(clusterName string, networkInterfaceAddress string, platform *baremetal.Platform, userDataSecret string) *baremetalprovider.BareMetalMachineProviderSpec {
+func provider(clusterName string, networkInterfaceAddress string, platform *baremetal.Platform, osImage string, userDataSecret string) *baremetalprovider.BareMetalMachineProviderSpec {
+	// The rhcos-downloader container launched by the baremetal-operator downloads the image,
+	// compresses it to speed up deployments and makes it available on provisioningIP via http
+	provisioningIP := platform.ClusterProvisioningIP
```
Comment (abhinavdahiya) [inline pkg/asset/machines/baremetal/machines.go]: nit: this is only used once, so directly using `platform.ClusterProvisioningIP` at call site might be better.

File: pkg/asset/machines/baremetal/machines.go [hunk 16]
```diff
@@ -60,11 +62,18 @@ func Machines(clusterID string, config *types.InstallConfig, pool *types.Machine
...
+	imageFilename := path.Base(osImage)
+	compressedImageFilename := strings.Replace(imageFilename, "openstack", "compressed", 1)
```
Comment (abhinavdahiya) [inline pkg/asset/machines/baremetal/machines.go]: can you add comment in terms of what API this is trying to follow??

File: pkg/asset/machines/baremetal/machines.go [hunk 17]
```diff
@@ -60,11 +62,18 @@ func Machines(clusterID string, config *types.InstallConfig, pool *types.Machine
...
+	cacheImageURL := fmt.Sprintf("http://%s:6180/images/%s/%s", provisioningIP, imageFilename, compressedImageFilename)
+	cacheChecksumURL := fmt.Sprintf("%s.md5sum", cacheImageURL)
 	return &baremetalprovider.BareMetalMachineProviderSpec{
 		Image: baremetalprovider.Image{
```
Comment (abhinavdahiya) [inline pkg/asset/machines/baremetal/machines.go]: how does it know what version of image to load? the URL seems like a URL to a directory?

Comment (abhinavdahiya) [reply pkg/asset/machines/baremetal/machines.go]: I might be wrong, therefore would love to see a comment here...

File: pkg/tfvars/baremetal/baremetal.go [hunk 18]
```diff
@@ -80,10 +83,16 @@ func TFVars(libvirtURI, ironicURI, bootstrapOSImage, externalBridge, provisionin
 		}
 
 		// Instance Info
+		// The rhcos-downloader container downloads the image, compresses it to speed up deployments
+		// and then makes it available on bootstrapProvisioningIP via http
+		imageFilename := path.Base(image)
```
Comment (abhinavdahiya) [inline pkg/tfvars/baremetal/baremetal.go]: Are we doing the same thing as Machine here for bootstrap? why not use the Machine object itself.

File: pkg/types/baremetal/platform.go [hunk 19]
```diff
@@ -33,10 +24,16 @@ type Platform struct {
 	// Default is qemu:///system
 	LibvirtURI string `json:"libvirtURI,omitempty"`
 
-	// IronicURI is the identifier for the Ironic connection.  It must be
-	// reachable from the host where the installer is run.
+	// ClusterProvisioningIP is the IP on the provisioning host private/provisioning
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go]: I don't follow what this field is doing??

`provisioning host` what is this? the host that is getting provisioned to join the cluster as master/compute ?

`ClusterProvisioningIP is the IP on the provisioning host private/provisioning network for the host running the installer` this sentence isn't making a whole lot of sense to me _ can we try to make this a little simpler for non-expert person like me_

File: pkg/types/baremetal/platform.go [hunk 20]
```diff
@@ -33,10 +24,16 @@ type Platform struct {
...
+	// network for the host running the installer, can be used to cache some
+	// downloaded content e.g RHCOS/IPA images
 	// +optional
-	IronicURI string `json:"ironicURI,omitempty"`
+	ClusterProvisioningIP string `json:"provisioningHostIP,omitempty"`
+
+	// BootstrapProvisioningIP is the IP used on the bootstrap VM to
+	// bring up provisioning services.
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go]: what are these provisioning services

File: pkg/types/baremetal/platform.go [hunk 21]
```diff
@@ -33,10 +24,16 @@ type Platform struct {
...
 	LibvirtURI string `json:"libvirtURI,omitempty"`
 
-	// IronicURI is the identifier for the Ironic connection.  It must be
-	// reachable from the host where the installer is run.
+	// ClusterProvisioningIP is the IP on the provisioning host private/provisioning
+	// network for the host running the installer, can be used to cache some
+	// downloaded content e.g RHCOS/IPA images
 	// +optional
-	IronicURI string `json:"ironicURI,omitempty"`
+	ClusterProvisioningIP string `json:"provisioningHostIP,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go]: doesn't provide any guidelines on if this IP can be part of the machine CIDR? or the pod CIDR etc..

File: pkg/types/baremetal/validation/platform.go [hunk 22]
```diff
@@ -24,8 +24,12 @@ func ValidatePlatform(p *baremetal.Platform, n *types.Networking, fldPath *field
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("libvirtURI"), p.LibvirtURI, err.Error()))
 	}
 
-	if err := validate.URI(p.IronicURI); err != nil {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("ironicURI"), p.LibvirtURI, err.Error()))
+	if err := validate.IP(p.ClusterProvisioningIP); err != nil {
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/validation/platform.go]: validate which range this can belong to. machine/pod/service..

File: pkg/asset/cluster/baremetal/baremetal.go (hunk 14 above)
Comment (stbenjam) [reply pkg/asset/cluster/baremetal/baremetal.go]: I can't think of a reason to keep it, will remove it. AFAIK this data is really only preserved for cluster destruction, but the name isn't really clear.  I've seen several people (including me in beginning) adding fields to the install-config repeatedly confused by this and try to add all the data here. Not sure if that's common for others but it could use a different name, or explict comments everywhere saying what it's for.

File: data/data/bootstrap/baremetal/systemd/units/ironic.service (hunk 13 above)
Comment (hardys) [reply data/data/bootstrap/baremetal/systemd/units/ironic.service]: Good point, I copied the file from the previous coredns integration and assumed it was required to ensure the service only started after the ignition pivot is complete, but that may be incorrect.

@celebdor perhaps you can clarify, was this needed for the coredns service, or just a copy/paste addition we can safely remove?

File: pkg/asset/cluster/baremetal/baremetal.go (hunk 14 above)
Comment (hardys) [reply pkg/asset/cluster/baremetal/baremetal.go]: Yeah it's probably not needed, it got changed here because IronicURI was already there, and I modified that existing name to BootstrapProvisioningIP via sed and some manual formatting fixes.

File: pkg/asset/machines/baremetal/machines.go (hunk 16 above)
Comment (hardys) [reply pkg/asset/machines/baremetal/machines.go]: Ack sure, it's basically to align with the rhcos downloader filename convention:

https://github.com/openshift/ironic-rhcos-downloader/blob/master/get-resource.sh#L25

That downloads the openstack qcow2, then compresses it to make baremetal deployment faster (maybe we can figure out some way to do that by default in future).

I'll add an example of the expected URL and a link to the rhcos-downloader container as a comment.

File: pkg/asset/machines/baremetal/machines.go (hunk 17 above)
Comment (hardys) [reply pkg/asset/machines/baremetal/machines.go]: Same as above, it's aligning with the file paths/URL generated via rhcos-downloader, so I'll add a comment that clarifies both the filename and the URL.

File: pkg/tfvars/baremetal/baremetal.go (hunk 18 above)
Comment (hardys) [reply pkg/tfvars/baremetal/baremetal.go]: It's not quite the same, the rhcos-downloader is running in a different location (on the bootstrap VM vs the cluster in the machine case), so we have to use a different IP.

May be this can be a bit DRY-er, but perhaps I can do that as a followup, with some URL generating utility function or something?

File: pkg/types/baremetal/platform.go (hunk 19 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: Ah yeah, the comment is garbled because of some more sed-replacement of a previous name, I'll fix this up.

To clarify, there are two places where the provisioning services (e.g ironic and dependencies) run, first on the bootstrap VM (accessible via BootstrapProvisioningIP to provision the master nodes via terraform), then later on the cluster managed via the Baremetal Operator, at this point the IP used changes to ClusterProvisioning IP.

Note that in both cases the network is a dedicated provisioning network, so unrelated to the normal pod networking and OCP controlplane (it's a dedicated second NIC for PXE boot etc)

File: pkg/types/baremetal/platform.go (hunk 20 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: Do we want to list them here?  It's supposed to be an implementation detail, but it's currently ironic and dependencies, e.g all-the-things started via the `ironic.service`

File: pkg/types/baremetal/platform.go (hunk 21 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: Ack, yeah it definitely cannot be part of those CIDRs so some validation would be a good idea.

I had hoped to change this to something like ProvisioningCidr in due course, but we've not quite got all the templating in place to enable that yet.

File: pkg/types/baremetal/validation/platform.go (hunk 22 above)
Comment (hardys) [reply pkg/types/baremetal/validation/platform.go]: Ack, it needs to be a dedicated additional network so it must not overlap

**Comment by abhinavdahiya** [comment]
/hold

The master is open only for bug fixes as we prepare for 4.2 release. this need a bugzilla targeted for 4.2 to merge.

File: pkg/types/baremetal/validation/platform.go (hunk 22 above)
Comment (hardys) [reply pkg/types/baremetal/validation/platform.go]: I raised https://github.com/openshift/installer/issues/2210 so this validation can perhaps be a follow-up PR?

File: pkg/types/baremetal/platform.go (hunk 21 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: I raised #2210 so this validation can perhaps be a follow-up PR?

**Comment by hardys** [comment]
Ok I addressed all comments except the validation of *ProvisioningIP, perhaps we can address that via a follow-up ref #2210 ?

**Comment by eparis** [comment]
This is ok to merge without a BZ if we believe it has met the typical engineering review bar for such a PR.
/hold cancel

File: pkg/types/baremetal/platform.go [hunk 23]
```diff
@@ -33,10 +24,18 @@ type Platform struct {
 	// Default is qemu:///system
 	LibvirtURI string `json:"libvirtURI,omitempty"`
 
-	// IronicURI is the identifier for the Ironic connection.  It must be
-	// reachable from the host where the installer is run.
+	// ClusterProvisioningIP is the IP on the dedicated provisioning network
+	// where the baremetal-operator pod runs provisioning services,
+	// and http server to cache some downloaded content e.g RHCOS/IPA images
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go]: nit: `and the`

File: pkg/types/baremetal/platform.go [hunk 24]
```diff
@@ -33,10 +24,18 @@ type Platform struct {
...
+	// These are used during the deployment of worker machines
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go]: nit, this is used for Machine objects. in case of DR.. even master machines should be recreate able.. so I think we should drop this from the API.

File: pkg/types/baremetal/platform.go [hunk 25]
```diff
@@ -33,10 +24,18 @@ type Platform struct {
...
 	// +optional
-	IronicURI string `json:"ironicURI,omitempty"`
+	ClusterProvisioningIP string `json:"provisioningHostIP,omitempty"`
+
+	// BootstrapProvisioningIP is the IP used on the bootstrap VM to
+	// bring up provisioning services that are used by terraform during the
```
Comment (abhinavdahiya) [inline pkg/types/baremetal/platform.go]: Can't have terraform in public API docs. `that are used to create the control-plane machines`

File: pkg/types/baremetal/platform.go (hunk 24 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: Ack yes it could also be used to rebuild master machines, will update

File: pkg/types/baremetal/platform.go (hunk 25 above)
Comment (hardys) [reply pkg/types/baremetal/platform.go]: ack will do

**Comment by hardys** [comment]
@abhinavdahiya thanks for the prompt review, latest comments addressed

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1036/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1037/
